### PR TITLE
[Ready for Review] Fixing creation of config.yaml and secret.yaml

### LINKED ISSUE #42: Docker ID parsed as null in deploy.sh container mode
When the BinderHub is deployed, we're getting internal server errors due to the Docker ID not being parsed correctly to `config.yaml`.

Plan is to replace the python scripts `create_config.py` and `create_secret.py` with `sed` commands in `deploy.sh`. Then test the DOCKER_ORG `if` statement.

### Update

Changing python scripts for sed commands worked, but docker username is still being parsed as `null` to `config.yaml`.

Required:
```
config:
  BinderHub:
    use_registry: true
    image_prefix: dockerid/binder-dev-
```

Returned:
```
config:
  BinderHub:
    use_registry: true
    image_prefix: null/binder-dev-
```

### PR DESCRIPTION
**Pull request by sgibson91**
In this PR:

* `create_config.py` and `create_secret.py` are deleted
* these scripts are replaced with `sed` commands in `deploy.sh`
* variable names are tidied in `deploy.sh`
* README documentation is updated
* placeholders in `config-template.yaml` and `secret-template.yaml` are standardised
* generated `*.log` files are git-ignored
* `deploy.sh` is now no longer dependent on python, therefore remove python install commands from `setup.sh`

fixes #42

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,6 @@ config.json
 # Ignore config and secret yaml files
 config.yaml
 secret.yaml
+
+# Ignore geerated log files
```
Comment (sgibson91) [inline .gitignore]:
```suggestion
# Ignore generated log files
```

**Comment by sgibson91** [comment]
@tmbgreaves I'm making a switch from python to `sed` commands. I've marked the lines in `setup.py` as comments that refer to python and I think should be deleted. Should we check for/install `sed` in `setup.py` as well?

**Comment by sgibson91** [comment]
Been running into some issues with tiller today:
```
--> Verify Client and Server are running the same version number:
Client: &version.Version{SemVer:"v2.14.0", GitCommit:"05811b84a3f93603dd6c2fcfe57944dfa7ab7fd0", GitTreeState:"clean"}
E0605 15:43:28.814412   73802 portforward.go:400] an error occurred forwarding 55746 -> 44134: error forwarding port 44134 to pod f9bdf58c6c52b1b5c8ede06b00ce5101c9a1fd2790f3b99c63dd783ad3deb2f2, uid : exit status 1: 2019/06/05 14:43:28 socat[8813] E connect(6, AF=2 127.0.0.1:44134, 16): Connection refused
E0605 15:43:29.940926   73802 portforward.go:400] an error occurred forwarding 55746 -> 44134: error forwarding port 44134 to pod f9bdf58c6c52b1b5c8ede06b00ce5101c9a1fd2790f3b99c63dd783ad3deb2f2, uid : container not running (f9bdf58c6c52b1b5c8ede06b00ce5101c9a1fd2790f3b99c63dd783ad3deb2f2)
E0605 15:43:31.702942   73802 portforward.go:400] an error occurred forwarding 55746 -> 44134: error forwarding port 44134 to pod f9bdf58c6c52b1b5c8ede06b00ce5101c9a1fd2790f3b99c63dd783ad3deb2f2, uid : container not running (f9bdf58c6c52b1b5c8ede06b00ce5101c9a1fd2790f3b99c63dd783ad3deb2f2)
E0605 15:43:34.297994   73802 portforward.go:400] an error occurred forwarding 55746 -> 44134: error forwarding port 44134 to pod f9bdf58c6c52b1b5c8ede06b00ce5101c9a1fd2790f3b99c63dd783ad3deb2f2, uid : container not running (f9bdf58c6c52b1b5c8ede06b00ce5101c9a1fd2790f3b99c63dd783ad3deb2f2)
E0605 15:44:08.294260   73802 portforward.go:340] error creating error stream for port 55746 -> 44134: Timeout occured
E0605 15:44:33.685023   73802 portforward.go:362] error creating forwarding stream for port 55746 -> 44134: Timeout occured
E0605 15:45:03.197795   73802 portforward.go:362] error creating forwarding stream for port 55746 -> 44134: Timeout occured
E0605 15:45:39.031531   73802 portforward.go:340] error creating error stream for port 55746 -> 44134: Timeout occured
E0605 15:46:41.150742   73802 portforward.go:362] error creating forwarding stream for port 55746 -> 44134: Timeout occured
E0605 15:48:02.291839   73802 portforward.go:362] error creating forwarding stream for port 55746 -> 44134: Timeout occured
Error: cannot connect to Tiller
```

**Comment by sgibson91** [comment]
BinderHub seemed to deploy fine after all of that.

File: deploy.sh [hunk 2]
```diff
@@ -252,28 +252,28 @@ helm repo update
...
+sed -e "s/<apiToken>/$apiToken/" \
+-e "s/<secretToken>/$secretToken/" \
+-e "s/<docker-id>/$DOCKER_USERNAME/" \
+-e "s/<password>/$DOCKER_PASSWORD/" \
+./secret-template.yaml > ./secret.yaml
 
-# Format name for kubernetes 
-HELM_BINDERHUB_NAME=$(echo ${BINDERHUB_NAME} | tr -cd '[:alnum:]-.' | tr '[:upper:]' '[:lower:]' | sed -re 's/^([.-]+)//' -re 's/([.-]+)$//' )
+# Format name for kubernetes
+HELM_BINDERHUB_NAME=$(echo ${BINDERHUB_NAME} | tr -cd '[:alnum:]-.' | tr '[:upper:]' '[:lower:]' | sed -e 's/^([.-]+)//' -e 's/([.-]+)$//' )
```
Comment (sgibson91) [inline deploy.sh]: Removed the `-r` flag from this line as that is not a valid `sed` option on OSX (meaning `deploy.sh` failed when run locally on a Mac). Does this expression need changing or is it ok?

Comment (tmbgreaves) [reply deploy.sh]: I think the fix here is to change '-r' to '-E'. -r is apparently not posix-compliant, where -E is, and we do need to switch on extended regexes here to make it do what we want. Please could you try -E on your mac and see if it works?

File: config-template.yaml [hunk 3]
```diff
@@ -2,3 +2,4 @@ config:
   BinderHub:
     use_registry: true
     image_prefix: <docker-id>/<prefix>-
+    hub_url: http://<jupyterhub-ip>
```
Comment (tmbgreaves) [inline config-template.yaml:5]:
```suggestion
    hub_url: http://<jupyterhub-ip>/
```
Does this need a trailing slash?

File: deploy.sh (hunk 2 above)
Comment (tmbgreaves) [inline deploy.sh]:
```suggestion
HELM_BINDERHUB_NAME=$(echo ${BINDERHUB_NAME} | tr -cd '[:alnum:]-.' | tr '[:upper:]' '[:lower:]' | sed -E -e 's/^([.-]+)//' -e 's/([.-]+)$//' )
```

I'd not realised -r wasn't POSIX-compliant -- oops! -E should do the same job and be POSIX.

UPDATED 2019-06-07-1426 to remove multiple -E

**Review by tmbgreaves — COMMENTED** [review]
As a general comment - I think I'd be in favour of just dropping the python install parts of the setup script, assuming there's no intention to bring back use of Python3 in the near future. If anyone wants it back they can always look back in commit history :-)

File: config-template.yaml (hunk 3 above)
Comment (sgibson91) [reply config-template.yaml:5]: I don't think so, I've never used one

**Comment by sgibson91** [comment]
> As a general comment - I think I'd be in favour of just dropping the python install parts of the setup script, assuming there's no intention to bring back use of Python3 in the near future. If anyone wants it back they can always look back in commit history :-)

I wrote the python stuff based on [nicain/binder-deploy](https://github.com/nicain/binder-deploy) and before I learned about sed, but I think the sed method is a lot cleaner. I can't guarantee I won't want to bring python back in the future as I don't quite know how I want to implement giving the user a choice between DockerHub and an Azure Container Registry yet. But yes, we can always scrape the commit history :)

File: deploy.sh (hunk 2 above)
Comment (sgibson91) [reply deploy.sh]: Testing with `-E` now

Comment (sgibson91) [reply deploy.sh]:
```
sed: 1: "E
": invalid command code E
```

Comment (sgibson91) [reply deploy.sh]: I think these are the only allowed options for OSX sed: https://ss64.com/osx/sed.html
We should try to be as general as possible.

**Comment by sgibson91** [comment]
@tmbgreaves Do you mind going over `setup.sh` and marking exactly which python commands are safe to delete? I did a search for python and commented all those lines, but we might need python to install Helm?

**Comment by tmbgreaves** [comment]
> @tmbgreaves Do you mind going over `setup.sh` and marking exactly which python commands are safe to delete?

Addressed in PR #46 along with a bunch of other setup bugs and features I discovered in the process of looking through the python setup :-)

I've run setup.sh on Fedora, Ubuntu, CentOS, Arch, and Debian, all of which completed without errors.

File: deploy.sh (hunk 2 above)
Comment (tmbgreaves) [reply deploy.sh]: Updated initial suggestion to single -E

**Pull request merged**